Permissions.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+
+import { NativeModules, PermissionsAndroid, Platform } from 'react-native';
+
+const { WebRTCModule } = NativeModules;
+
+/**
+ * Class implementing a subset of W3C's Permissions API as defined by:
+ * https://www.w3.org/TR/permissions/
+ */
+class Permissions {
+    /**
+     * Possible result values for {@link query}, in accordance with:
+     * https://www.w3.org/TR/permissions/#status-of-a-permission
+     */
+    RESULT = {
+        DENIED: 'denied',
+        GRANTED: 'granted',
+        PROMPT: 'prompt'
+    };
+
+    /**
+     * This implementation only supports requesting these permissions, a subset
+     * of: https://www.w3.org/TR/permissions/#permission-registry
+     */
+    VALID_PERMISSIONS = [ 'camera', 'microphone' ];
+
+    _lastReq = Promise.resolve();
+
+    /**
+     * Helper for requesting Android permissions. On Android only one permission
+     * can be requested at a time (unless the multi-permission API is used,
+     * but we are not using that for symmetry with the W3C API for querying)
+     * so we'll queue them up.
+     */
+    _requestPermissionAndroid(perm) {
+        return new Promise((resolve, reject) => {
+            PermissionsAndroid.request(perm).then(
+                granted => resolve(granted === PermissionsAndroid.RESULTS.GRANTED),
+                () => resolve(false));
+        });
+    }
+
+    /**
+     * Validates the given permission descriptor.
+     */
+    _validatePermissionDescriptior(permissionDesc) {
+        if (typeof permissionDesc !== "object") {
+            throw new TypeError("Argument 1 of Permissions.query is not an object.");
+        }
+        if (typeof permissionDesc.name === "undefined") {
+            throw new TypeError("Missing required 'name' member of PermissionDescriptor.");
+        }
+        if (this.VALID_PERMISSIONS.indexOf(permissionDesc.name) === -1) {
+            throw new TypeError("'name' member of PermissionDescriptor is not a valid value for enumeration PermissionName.");
+        }
+    }
+
+    /**
+     * Method for querying the status of a permission, according to:
+     * https://www.w3.org/TR/permissions/#permissions-interface
+     */
+    query(permissionDesc) {
```

Comment:
can you describe the type for permissionDesc and then use that type in all methods in this file which use it ? it's not easy to find the exact structure in the link above.

Reply:
👍